### PR TITLE
Table Panel: Conditional Formatting POC

### LINKED ISSUE #11418: [Feat] Conditional formatting
## Introduction

As of writing (Grafana v5.1-pre) we have support for 
* coloring background/value in the Singlestat panel based on a numeric range and 3 colors
* coloring cell/value/row in the Table panel based on a numeric range and 3 colors

image

image

There are a lot of feature requests/issues requesting greater support of coloring in table panel and the singlestat panel:
* #7631 [Feature request] Allow Text to Value mappings in Singlestat panel
* #3601 table panel: colorize cell value based on regex pattern match
* #9500 Feature request: Allow thresholds for Time column in tables
* #3608 Thresholds with more than 3 comma-separated values
* #8404 [Feature request] Singlestat null value color 
* #4821 Table Panel coloring shouldn't require display of referenced column
* #9709 Single stat panel with gradient color
* #7411 [Feature request] Variable threshold values in Singlestats panel
* #7286 [Feature request] Row and/or Column styles in Table

Historically we've received a few PR's that tried to solve some of the issues above, but we haven't been able to merge them because we feel the solutions are either hard to understand from a user perspective or either it makes the Grafana code base/features more and more shattered.

## Proposal

This issue aims to find one solution for solving all (or most) of the issues above and by that delivering a good and consistent user experience. By that said, we probably want to split this up in smaller issues if we decide that this is a good way forward.

In general the *Conditions editor* for alerts are very similar to a rule/conditional formatting editor functionality and would probably fit quite nice with a few modifications. This is something to take into considerations when implementing this.

image

### Conditional rules evaluator

A panel needing conditional formatting shouldn't need to implement their own solution for evaluating conditional rules thus making this a core component/helper library is a must. 
It would be nice if this component also could return a list of rule types that it supports.

Examples of rule types:
* Value is empty
* Value is not empty
* Text contains
* Text does not contain
* Text starts with
* Text ends with
* Text is exactly
* Date is
* Date is before
* Date is after
* Greater than
* Greater than or equal to
* Less than
* Less than or equal to
* Is equal to
* Is not equal to
* Is between
* Is not between

### Singlestat panel - colorize background/value

Changes here compared to current functionality is that 
* it allow you to define unlimited amount of color formatting's compared to today's maximum of three.
* it allow support for coloring based on empty/non-empty values
* it allow support for coloring based on string/text values
* it allow support for coloring based on date/duration values

#### UX mockup/example to underline functionality needed/wanted

image

### Table panel - colorize cells/values/rows

Changes here compared to current functionality is that 
* it allow you to define unlimited amount of color formatting's compared to today's maximum of three.
* it allow support for coloring based on empty/non-empty values
* it allow support for coloring based on string/text values
* it allow support for coloring based on date/duration values
* it allow support for coloring a row based on a hidden/non-rendered column

Currently you can define thresholds for each defined *Column style*. Maybe we want to continue having this possibility since we can make smart decisions of what columns to apply the conditional formatting rules to. 

On the other hand it quite weird to continue using the column styles since they're basically provides options for how to render column cells/convert raw column value using specific data types and options etc. Regarding coloring we currently supports cell/value/row which affects cells or rows and rules are evaluated based on  the raw data returned from data source query, that is not after conversion/options. In the mockup below there's a really simple UX which basically would scan all rows and columns returned from data source query to apply conditional formatting.

One option to having conditional formatting inside the *Column style* tab would be to allow conditional formatting rules to be applied per columns basis under a specific tab. Basically like the mockup below but with additional logic to decide which columns the rules should be applied to. This solution have an advantage because it can support conditional formatting of rows based on a hidden/non-rendered column.

#### UX mockup/example to underline functionality needed/wanted

image

### Forward compatibility

Any changes needed to existing Singlestat and Table panel JSON models should be forward compatible, meaning that existing saved dashboards with Singlestat/Table panel having coloring/threshold enabled should be migrated/converted to the new solution when such a dashboard is rendered.

### PR DESCRIPTION
**Pull request by codeincarnate**
**What is this feature?**

This draft PR implements a first stab at conditional formatting for the table panel. Currently, it uses thresholds to apply colors to rows if a numeric value of a field in that row surpasses a given threshold. At present there are numerous details left to figure out including but not limited to:

- What to do when multiple numeric fields are present
- If it makes sense to apply formatting to both columns and rows, or simply to do so for one or the other
- If this meshes well with field overrides
- Figuring out appropriate text coloring based on theme and applied threshold color
- What to do in the case of nested tables
- How to properly handle percentage thresholds, not just absolute thresholds
- If base thresholds should ever be applied or if an option should be provided for this (currently base threshold is ignored)
- What to do for hovered rows
- Figuring out styling such that this looks good, and is up to the visual standard that Grafana visualizations provide

Here's a short video of this POC in action:

https://github.com/grafana/grafana/assets/199847/d7162a6f-e428-4de6-ad8a-e6f3bdf7c756

**Why do we need this feature?**

Conditional formatting by row is one of the most requested features in Grafana. It's desirable for a wide variety of use cases including observability, monitoring, analytics, business intelligence, and many many (many) more. In addition to being one of the oldest open issues against Grafana, we've had requests for this feature from all quarters.

**Who is this feature for?**

Most if not all of the users of the table panel users in Grafana

**Which issue(s) does this PR fix?**:

Fixes #11418

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by codeincarnate** [comment]
Thinking a bit more about this, going to avoid highlighting columns. There are certain areas where there are a lot of columns, though it seems like pivoting to row in those cases would make more sense. Being that rows are the more common use case here going to focus on that first.

**Comment by gtk-grafana** [comment]
AWESOME.
My only question is if there will be a way for calling component to define row colors based on string/enum values instead of simply numeric logic. Doesn't need to be exposed in the panel options, but if we can use the log level to color the table row that would solve a huge headache for logs users of the table

**Comment by codeincarnate** [comment]
@gtk-grafana that's an interesting question! Actually we probably do need something like this regardless as a common use case is to have an enumerated set of strings map to formats. i.e. Good would be green, Bad would be red, etc. Something to think about in moving this forward.

**Comment by codeincarnate** [comment]
@torkelo wanted to bring this to your attention quickly. It's still extremely early stage but thought that you would likely have some thoughts around implementation. Also considering how this might be achieved with field overrides (and collecting this all in a design doc as well) 😄

**Comment by torkelo** [comment]
@codeincarnate will take a look tomorrow

File: packages/grafana-ui/src/components/Table/Table.tsx [hunk 1]
```diff
@@ -79,6 +83,67 @@ export const Table = memo((props: Props) => {
...
+        // to consider the case of multiple numeric fields, may 
+        // need to use field overrides but 
+        // not sure about the feasability of that approach
+        if (value >= threshold.value && threshold.value !== (-Infinity)) {
+          if (colorMap[i] === undefined) {
+            colorMap[i] = threshold.color;
+            break;
+          }
+        }  
+      }
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Table/Table.tsx:120]: @codeincarnate 

Why just thresholds?? we have many colors schemes in grafana. I am really confusing by this . 

The problem seems pretty simple. Have the color of one of the fields apply to the row. Why introducing all this custom rowToColor and thresholds logic??

https://github.com/grafana/grafana/pull/83247

Comment (codeincarnate) [reply packages/grafana-ui/src/components/Table/Table.tsx:120]: In terms of the thresholds, it's been one of the use cases that's been asked for (coloring beyond a given threshold). It's an initial stab at it, and wasn't sure about this route honestly :)

Have a design doc fleshing out more (including what you mentioned) here https://docs.google.com/document/d/1en6BS3V9qrjmo6sVQYWLdZQibGSaLWSnYSC_Y779Mrw/edit

**Comment by codeincarnate** [comment]
Closing in favor of https://github.com/grafana/grafana/pull/83939

**Pull request closed without merging**